### PR TITLE
1008 ideal integration online payment provider

### PR DESCRIPTION
**Pull request by yksflip**
This PR adds the online payment provider _mollie_ as a plugin. User can add credit to their ordergroups balance via various payment methods. wvengen implemented this already in the foodsoft-adams fork. I rewrote some parts to make it work with the newest mollie api version and added some bits to improve the usability.

Changes relevant to the foodsoft core:
* added new attributes to the financial_transaction model to keep track of the state of a transaction
* financial_transaction.value may be `nil` now, this indicates an incomplete transaction

Changes on the plugin:
* payment_method is no longer choosen in the foodsoft form but on the mollie checkout
* information regarding the transaction fee is displayed 
* settings can be configured in the admin screen
* the transaction_type for mollie transactions can be configured in the administration settings

still open
* ~~[ ] move [accept_return_to](https://github.com/foodcoops/foodsoft/pull/1031/files#diff-5856be838b4186db9c017a6d2a3211e100851ae2c5418e4f52f29ddeec70e353R111) to ApplicationController?~~
* [x] open currency refactoring issue, implement FoodsoftMollie.currency
* [x] write financial transaction documentation
* [x] reverting a pending financial transaction?
* [x] handle amount < min ?
* [x] squash commits

### CODE REVIEW

**Review by wvengen — COMMENTED** [review]
Wow, great to see this coming back, now properly as a core Foodsoft plugin! :)

It's been a long time, probably some questions I'm asking are about pieces you've just copied from the original code. Nevertheless, it's good to remain critical (also on my past self).

There are two possibly important assumptions that change:
- A FinancialTransaction's `amount` may be `nil`
- A FinancialTransaction is mutable

This is something important to think through. Solid accounting has no changing history, but in this case we make an exception. I think this is justified, but I'm not 100% sure - both in Foodsoft core (not all parts of the code are tested fully), and by plugins/integrations people have written. This even affects the API (financial transations), where software using it may not expect `nil` for amount.
I think before this PR is merged, someone needs to have a good look at this. Perhaps you've done that already.

Do we need to make a new API version for this, hide the unfinished transactions by default, or accept a possibly incompatible change (the API is relatively young, so maybe ok)?
One way to get an answer to this is to ask this question on the forum and/or Github.

I also asked some questions to which I could have found the answer myself. Hope that's ok, and you're patient enough to go look, think about it, and respond.

Some general things to improve could be (partly already mentioned in separate comments):
- Use strong parameters, they're standard now, and with finance-related code it's good to be safe.
- I think it's good to document what we expect from financial transactions, just some background in the main Foodsoft (I guess in-tree) developer docs, about the assumptions. Maybe it's not really a question for you. I think it would have best been done already with the introduction of financial transaction types and such. But this PR touches something quite important, so I think writing down something about financial transactions may help even getting your own mental model right.

Please feel free to disagree too, I hope we can have a constructive conversation :)
Thanks for all your work in making this ready for prime-time!

**Comment by wvengen** [comment]
p.s. if the `amount` can be `nil` in the API, the API docs (swagger) may need updating.

File: db/migrate/20230915093041_add_payment_to_financial_transaction.rb
Comment (yksflip) [reply db/migrate/20230915093041_add_payment_to_financial_transaction.rb:21]: Yes you're right. I think for the mollie usecase we actually don't need to save this information, as it's not used anywhere. And if we wan't to it could still be fetched from the api...

File: plugins/mollie/README.md
Comment (yksflip) [reply plugins/mollie/README.md:21]: I think this is quite tricky. There is a way to get fee's from the mollie api, but e.g. paypal charges additional fees ... But maybe this is only the exception. 

Maybe this would be a sane config interface: 
* [] Charge ordergroup mollie fee's 
  * [] defaults from mollie-api
  * Input custom fee for  $payment_method

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (yksflip) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb:115]: tbh this is a copy pasta of your code :o
but I'll open a refactoring issue :)

File: plugins/mollie/README.md
Comment (yksflip) [reply plugins/mollie/README.md:21]: I thought a override might be useful if the actual costs differ for some reason (I've read smth about paypal having extra fee's, but I'm not sure). But yeah maybe overkill for the moment.

Via the [/methods endpoint](https://docs.mollie.com/reference/v2/methods-api/get-method#example) it is possible to get mollies fees. And yess it's a bit more complex :o
Though I haven't found a way to get a computed fee from mollie payment result.

Having the option to topup the fee amount would be great. I think I'll add the method and issuer selection again, so the fee can be computed beforehand.

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (yksflip) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb]: I choose to delete the return_to functions completely, afaik we don't need it atm.

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (yksflip) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb]: whoops, it was actually the opposite,  hope its more clear now?

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (yksflip) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb]: I guess this would be a big refactoring ... I'd open a issue and reference it in the according places. I could also make it call smth like `FoodsoftMollie.currency`, than later only one place needs to be changed

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (yksflip) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb]: I have no idea, I'm really not deep enough into RoR yet. I don't even know how to figure that out ... Can you give me another hint? :)

File: plugins/mollie/app/overrides/admin/configs/_tab_others/add_config.html.haml.deface
Comment (yksflip) [reply plugins/mollie/app/overrides/admin/configs/_tab_others/add_config.html.haml.deface:6]: yes, maybe that could be the case. This is the way vokoosterwold would like to use it, but there might be other use cases. I think once they appear we should try to add them. 
I'll try to learn more about the usage of ft types

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (wvengen) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb]: What about making this a Mollie configuration option (with a default to `EUR`), explaining that this needs to match Foodsoft's currency?
Available currencies at: https://docs.mollie.com/payments/multicurrency

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (wvengen) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb]: Ah, that makes more sense!

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (wvengen) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb]: See `set_currency` and `set_time_zone` in the ApplicationController.

An around hook would probably be best, setting it early in the request/response process, and clearing it afterwards.
Another option would be to set it like here, but then it needs to be set _always_ (maybe even when the plugin is disabled?), to be sure it doesn't look. An around hook is safer.

File: doc/API.md
Comment (wvengen) [inline doc/API.md:25]: Unrelated change, though useful :)

**Review by wvengen — COMMENTED** [review]
This is shaping up nicely!

**Comment by yksflip** [comment]
> This is shaping up nicely!

thx! your comment are very helpful :)

File: doc/API.md
Comment (yksflip) [reply doc/API.md:25]: ahh couldn't resist, I'll put it in a separate commit.

**Comment by yksflip** [comment]
Hey I've been working a bit more an introduced some changes:

  * transactions with nil value are no longer revertable
  * display payment_state in transaction views
  * strong params on api controller
  * api-key is now used in function call (around filter didn't work)
  * introduce currency config
    Mollie returns fee amounts always in EUR, no matter what currency is
    selected. Because of that there is a constraint to charge fees for
    EUR only.
  * introduce default amount config
  * some code refactoring

Hopefully I can finish up on this until end of this week :)

File: app/helpers/application_helper.rb
Comment (yksflip) [reply app/helpers/application_helper.rb]: I thought it would be the best to just display the payment state as a seperate column. It felt weird to use the amount column for this information.

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (yksflip) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb]: I'd go with silently assuming the _min_. Still users have to confirm the amount again on Mollie page.

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (yksflip) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb]: I had to realise that mollies method endpoint will always return the fee amount in EUR. I think for now the most straight forward thing is to make 'EUR' a constraint for the fee feature (I'd guess that's the most common case). If we find a way to retrieve the actual fee calculated by mollie we can move on to that. (Still calculating the right fee to increase the amount for different currencies before paying would be difficult)

File: plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb
Comment (yksflip) [reply plugins/mollie/app/controllers/payments/mollie_ideal_controller.rb]: I don't know why, but I've got 401 unauthorized errors when trying to this as a around hook. I've decided to just pass the api-key in the function calls directly (it's only twice)

**Comment by wvengen** [comment]
Please let me know when it is ready for review (comment or the re-request review button), as it's been a while. Would be great to have this included in the next release.

**Comment by yksflip** [comment]
hey @wvengen, thanks for the reminder. I know it's been a while ... I tried to write some documentation, though I don't feel very confident with it. I still don't understand very much of the banking feature. I see this docs as a starting point...
I'd happy to hear your feedback and would be great to see this finally merged :)

File: app/controllers/api/v1/financial_transactions_controller.rb
Comment (wvengen) [inline app/controllers/api/v1/financial_transactions_controller.rb]: What about moving the `include_complete` logic to `search_scope` in this class?

File: app/controllers/api/v1/user/financial_transactions_controller.rb
Comment (wvengen) [inline app/controllers/api/v1/user/financial_transactions_controller.rb]: idem

File: app/helpers/application_helper.rb
Comment (wvengen) [inline app/helpers/application_helper.rb:24]: I think this works, but is a little strange: if amount is not numeric, then `class_name` is actually undefined.

File: app/views/finance/financial_transactions/_transactions.html.haml
Comment (wvengen) [inline app/views/finance/financial_transactions/_transactions.html.haml:27]: Is the table still readable with a number of added columns (with possibly long names)?
This is especially relevant for foodcoops not using a payment plugin.

File: spec/swagger_helper.rb
Comment (wvengen) [inline spec/swagger_helper.rb:190]: typo: "a incomplete" -> "an incomplete"

File: app/models/financial_transaction.rb
Comment (wvengen) [inline app/models/financial_transaction.rb]: Just checking in if this todo is something that needs to be checked.

File: app/models/financial_transaction.rb
Comment (wvengen) [inline app/models/financial_transaction.rb:48]: Not sure how good the testing of reverting transactions is, ideally we'd test this.
But as we also want to get this PR merged, feel free to skip if this feature has no testing at all.

**Review by wvengen — COMMENTED** [review]
Thank you for working on this PR!
Some small comments, I think it would be almost ready for merging.

Super you added some documentation. I find the financial transaction types and such also not that easy to comprehend. I think it's mostly the Austrian foodcoops using this, so they can best provide input here. See also [this forum thread](https://forum.foodcoops.net/t/transactionclasses-and-types/374), here is an issue [showing a use-case](https://github.com/foodcoops/foodsoft/issues/781), and [this issue](https://github.com/foodcoops/foodsoft/issues/170) has also some background.

File: app/helpers/application_helper.rb
Comment (yksflip) [reply app/helpers/application_helper.rb:24]: should we set class_name to a default value then?

**Comment by yksflip** [comment]
thanks for your feedback!
I've added some basic model tests for financial_transactions, refactored the search_scope, and implemented the around_save.
I also just had a chat with leo and he mentioned the user documentation on https://docs.foodcoops.net/de/documentation/admin/finances

File: spec/models/financial_transaction_spec.rb
Comment (wvengen) [inline spec/models/financial_transaction_spec.rb:10]: you can use `let!` here if you want to make sure these things are created

File: spec/models/financial_transaction_spec.rb
Comment (wvengen) [inline spec/models/financial_transaction_spec.rb:28]: I'd suggest to create a financial transaction trait in the factory that makes it more expressive, and will allow it to function when we'd change behaviour (e.g. have an explicit _pending_ attribute or so).

```ruby
context "with a pending transaction" do
  let!(:ft) { create(:financial_transaction, :pending, ordergroup: ordergroup, amount: 20) }
  it "fails on revert" do
      expect { ft.revert!(ft.user) }.to raise_error(RuntimeError, 'Pending Transaction cannot be reverted')
  end
end
```

File: app/views/finance/financial_transactions/_transactions.html.haml
Comment (wvengen) [reply app/views/finance/financial_transactions/_transactions.html.haml:27]: If we don't need a config option, let's avoid it (as it's complex enough as it is). So if they're a straightforward way to derive this, yes. Looking at payment_plugin config seems sensible here, yes.

Could there be a situation where this information is still relevant but the payment plugin is disabled? E.g. a payment plugin was enabled, but now disabled? I think we can just hide it, because any payment-plugin-related things will fail anyway. But just checking. E.g. at some places we check if there's more than one financial transaction type, and only show certain things if there is (to avoid UI complexity when a feature is not used).

File: app/helpers/application_helper.rb
Comment (wvengen) [reply app/helpers/application_helper.rb:24]: Is it about `amount` being `nil`? In that case, we can just omit the `span` entirely and return `nil` directly.

**Review by wvengen — COMMENTED** [review]
Thanks for the changes! We're definitely getting there :rocket:
Great you found the documentation, I didn't know about this, thanks for sharing.

**Comment by yksflip** [comment]
how about this? https://github.com/foodcoops/foodsoft/pull/1031/files#diff-f5c62182c669e1defeb29be1ca2beac61b02af50970b7b8b823708aa855221f7R23

File: plugins/mollie/README.md
Comment (wvengen) [inline plugins/mollie/README.md:20]: What does "is sub" mean? (could it be that it used to be "is substracted from ...").
Also, "it is not set by default" -> "it is disabled by default" sounds clearer.

File: plugins/mollie/config/locales/nl.yml
Comment (wvengen) [inline plugins/mollie/config/locales/nl.yml:26]: Please use the standard way to format currencies, as the currency may not be €.
Ok, later I read that Mollie (currently) assumes EUR (is that still the case?), so maybe not important. But the english language does not have the currency, so good to bring in harmony.

**Review by wvengen — COMMENTED** [review]
Super! Some small notes, otherwise good to go, I think :rocket:

File: app/views/finance/financial_transactions/_transactions.html.haml
Comment (wvengen) [reply app/views/finance/financial_transactions/_transactions.html.haml:23]: Wait ... somehow I must have looked at an old commit, or misread it. You're already doing this! :)

File: plugins/mollie/config/locales/nl.yml
Comment (yksflip) [reply plugins/mollie/config/locales/nl.yml]: oh yes, thats now taken from the payment_currency field, just forgot to adapt the nl translation. 
The support for automatic fee calculation is currently only supported for EUR, because of api limitation

File: app/views/finance/index.html.haml
Comment (yksflip) [reply app/views/finance/index.html.haml:43]: yeah that wasn't very clever :D

File: app/views/home/index.html.haml
Comment (yksflip) [reply app/views/home/index.html.haml]: yes, I now reuse the already loaded financial transactions

File: app/views/finance/financial_transactions/_transactions.html.haml
Comment (yksflip) [reply app/views/finance/financial_transactions/_transactions.html.haml:58]: yes, I introduced a helper variable and use the same for th and td now

**Comment by yksflip** [comment]
@wvengen let's merge? :)

**Review by wvengen — APPROVED** [review]
Yes! Many many thanks for completing this (and for keeping up with my feedback, requesting changes several times).

Feel free to press the merge button! (And I would suggest rebase merge in this case - and I would reword the commits to add "PR #1031" but maybe that's overthinking it.)

**Comment by wvengen** [comment]
Ah, and for a future PR, it's good to mention the issue number in the PR _description_ (e.g. in this case "Implements #1008"), so that Github recognizes this and provides cross-references. The "1008" in the PR title didn't do this.

**Comment by yksflip** [comment]
Thank you for also keeping up and your helpful advise :)

**Pull request merged**
